### PR TITLE
Refactor/overlay components

### LINKED ISSUE #664: Update popover according to new overlays
### Issue type

**I'm submitting a ...**  (check one with "x")

* [ ] bug report
* [x] feature request

### Issue description

### LINKED ISSUE #668: Update context menu according to new overlays
### Issue type

**I'm submitting a ...**  (check one with "x")

* [ ] bug report
* [x] feature request

### Issue description

### LINKED ISSUE #683: Update search according to new overlays
### Issue type

**I'm submitting a ...**  (check one with "x")

* [ ] bug report
* [x] feature request

### PR DESCRIPTION
**Pull request by Tibing**
### Please read and mark the following check list before creating a pull request:

 - [x] I read and followed the [CONTRIBUTING.md](https://github.com/akveo/nebular/blob/master/CONTRIBUTING.md) guide.
 - [x] I read and followed the [New Feature Checklist](https://github.com/akveo/nebular/blob/master/DEV_DOCS.md#new-feature-checklist) guide.
 
 #### Short description of what this resolves:

Refactor popover, context-menu, and searches, implement with overlays.
Remove capability add something to layout top. All dynamic layout components have to work with overlays now.

BREAKING CHANGE:
`appendToLayoutTop` and `clearLayoutTop` methods was removed from `NbThemeService`. Instead of this methods, you have to use `NbOverlayService`. It's the extension of @angular/cdk overlays, so, check [documentation](https://material.angular.io/cdk/overlay/overview) first of all. 

Basic usage of overlays may look like this:

```ts
constructor(protected overlay: NbOverlayService) {
}

const overlayRef = overlay.create();
const overlayComponentPortal = new ComponentPortal(MyOverlayComponent);
overlayRef.attach(overlayComponentPortal);
```

Closes #683, closes  #664, closes #668.

### CODE REVIEW

File: src/framework/theme/components/layout/layout.component.ts [hunk 1]
```diff
@@ -337,18 +336,18 @@ export class NbLayoutComponent implements AfterViewInit, OnDestroy {
...
     protected renderer: Renderer2,
-    protected router: Router,
     @Inject(NB_WINDOW) protected window,
     @Inject(NB_DOCUMENT) protected document,
     @Inject(PLATFORM_ID) protected platformId: Object,
     protected layoutDirectionService: NbLayoutDirectionService,
     protected scrollService: NbLayoutScrollService,
     protected rulerService: NbLayoutRulerService,
     protected scrollTop: NbRestoreScrollTopHelper,
+    @Optional() protected overlayContainer: NbOverlayContainerAdapter,
```
Comment (yggg) [inline src/framework/theme/components/layout/layout.component.ts]: ~@Optional()~

File: src/framework/theme/components/cdk/overlay/overlay-trigger.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+ * not on the host or container.
+ * */
+export class NbClickTriggerStrategy extends NbTriggerStrategy {
+  protected show: Subject<Event> = new Subject();
+  readonly show$: Observable<Event> = this.show.asObservable();
+
+  protected hide: Subject<Event> = new Subject();
+  readonly hide$: Observable<Event> = observableMerge(
+    this.hide.asObservable(),
+    observableFromEvent<Event>(this.document, 'click')
```
Comment (yggg) [inline src/framework/theme/components/cdk/overlay/overlay-trigger.ts]: renderer.listen?

File: src/framework/theme/components/cdk/overlay/overlay-trigger.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+      .pipe(filter((event: Event) => this.isNotHostOrContainer(event))),
+  );
+
+  constructor(protected document: Document, protected host: HTMLElement, protected container: () => ComponentRef<any>) {
+    super(document, host, container);
+    this.subscribeOnHostClick();
+  }
+
+  protected subscribeOnHostClick() {
+    observableFromEvent(this.host, 'click')
```
Comment (yggg) [inline src/framework/theme/components/cdk/overlay/overlay-trigger.ts]: renderer.listen?

File: src/framework/theme/components/cdk/overlay/overlay-trigger.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+/**
+ * Creates open and close events streams, the toggle is empty.
+ * Fires open event when a mouse hovers over the host element and stay over at least 100 milliseconds.
+ * Fires close event when the mouse leaves the host element and stops out of the host and popover container.
+ * */
+export class NbHoverTriggerStrategy extends NbTriggerStrategy {
+
+  show$: Observable<Event> = observableFromEvent<Event>(this.host, 'mouseenter')
```
Comment (yggg) [inline src/framework/theme/components/cdk/overlay/overlay-trigger.ts]: renderer.listen?

File: src/framework/theme/components/cdk/overlay/overlay-trigger.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    .pipe(
+      delay(100),
+      takeUntil(observableFromEvent(this.host, 'mouseleave')),
```
Comment (yggg) [inline src/framework/theme/components/cdk/overlay/overlay-trigger.ts]: renderer.listen?

File: src/framework/theme/components/cdk/overlay/overlay-trigger.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+      repeat(),
+    );
+
+  hide$: Observable<Event> = observableFromEvent<Event>(this.host, 'mouseleave')
```
Comment (yggg) [inline src/framework/theme/components/cdk/overlay/overlay-trigger.ts]: renderer.listen?

File: src/framework/theme/components/cdk/overlay/overlay-trigger.ts [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    .pipe(
+      switchMap(() => observableFromEvent<Event>(this.document, 'mousemove')
```
Comment (yggg) [inline src/framework/theme/components/cdk/overlay/overlay-trigger.ts]: renderer.listen?

File: src/framework/theme/components/cdk/overlay/overlay-trigger.ts [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    );
+}
+
+/**
+ * Creates open and close events streams, the toggle is empty.
+ * Fires open event when a mouse hovers over the host element and stay over at least 100 milliseconds.
+ * Fires close event when the mouse leaves the host element.
+ * */
+export class NbHintTriggerStrategy extends NbTriggerStrategy {
+  show$: Observable<Event> = observableFromEvent<Event>(this.host, 'mouseenter')
```
Comment (yggg) [inline src/framework/theme/components/cdk/overlay/overlay-trigger.ts]: renderer.listen?

File: src/framework/theme/components/cdk/overlay/overlay-trigger.ts [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    .pipe(
+      delay(100),
+      takeUntil(observableFromEvent(this.host, 'mouseleave')),
```
Comment (yggg) [inline src/framework/theme/components/cdk/overlay/overlay-trigger.ts]: renderer.listen?

File: src/framework/theme/components/cdk/overlay/overlay-trigger.ts [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+      repeat(),
+    );
+
+  hide$: Observable<Event> = observableFromEvent(this.host, 'mouseleave');
```
Comment (yggg) [inline src/framework/theme/components/cdk/overlay/overlay-trigger.ts]: renderer.listen?

File: src/framework/theme/components/context-menu/context-menu.directive.ts [hunk 11]
```diff
@@ -63,110 +68,116 @@ import { NbPlacementHelper } from '../popover/helpers/placement.helper';
...
-    this.menuTag = tag;
-    this.popover.context = Object.assign(this.context, { tag });
-  }
+  tag: string;
+
+  /**
+   * Basic menu items, will be passed to the internal NbMenuComponent.
+   * */
+  @Input('nbContextMenu')
+  set _items(items: NbMenuItem[]) {
```
Comment (nnixaa) [inline src/framework/theme/components/context-menu/context-menu.directive.ts]: `_`

File: src/playground/popover/popover-custom-component.component.ts [hunk 12]
```diff
@@ -10,6 +10,9 @@ import { NbDynamicToAddComponent } from '../shared/dynamic.component';
 @Component({
   selector: 'nb-popover-custom-component',
   templateUrl: './popover-custom-component.component.html',
+  styles: [`nb-layout-column {
```
Comment (nnixaa) [inline src/playground/popover/popover-custom-component.component.ts]: new lines

File: src/playground/popover/popover-showcase.component.ts [hunk 13]
```diff
@@ -10,6 +10,10 @@ import { Component } from '@angular/core';
 @Component({
   selector: 'nb-popover-showcase',
   templateUrl: './popover-showcase.component.html',
+  styles: [`:host {
```
Comment (nnixaa) [inline src/playground/popover/popover-showcase.component.ts]: new lines

**Pull request merged**